sourcetool/pkg/auth/implementation.go

```diff
@@ -0,0 +1,191 @@
+	if err := json.Unmarshal(body, &tokenResp); err != nil {
+		return "", fmt.Errorf("failed to parse token response: %w", err)
+	}
+
+	if tokenResp.Error != "" {
+		return "", fmt.Errorf("%s", tokenResp.Error)
+	}
+
+	if tokenResp.AccessToken == "" {
+		return "", fmt.Errorf("no access token received")
+	}
+
+	return tokenResp.AccessToken, nil
+}
+
+// openBrowser shells out to open the system's browser to load the
+// device authorization webpage.
+func (di *defaultImplementation) openBrowser(authURL string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default:
+		cmd = "xdg-open"
+	}
+
+	return exec.Command(cmd, append(args, authURL)...).Start() //nolint:gosec // yes variable input
+}
+
+// requestDeviceCode requests a device code from GitHub
+func (di *defaultImplementation) requestDeviceCode(ctx context.Context) (*DeviceCodeResponse, error) {
+	data := url.Values{}
+	data.Set("client_id", oauthClientID)
+	data.Set("scope", strings.Join(oauthScopes, " "))
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, deviceCodeURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create device code request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	httpClient := &http.Client{Timeout: 30 * time.Second}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to request device code: %w", err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("device code request failed with status: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading device code response: %w", err)
+	}
+
+	var deviceResp DeviceCodeResponse
+	if err := json.Unmarshal(body, &deviceResp); err != nil {
+		return nil, fmt.Errorf("parsing device code response: %w", err)
+	}
+
+	return &deviceResp, nil
+}
+
+// persistToken stores the token into the system config directory
+func (di *defaultImplementation) persistToken(token string) error {
+	dir, err := os.UserConfigDir()
+	if err != nil {
+		return fmt.Errorf("getting user config dir: %w", err)
+	}
+
+	if err := os.MkdirAll(filepath.Join(dir, configDirName), os.FileMode(0o700)); err != nil {
+		return fmt.Errorf("creating config directory: %w", err)
+	}
+
+	err = os.WriteFile(filepath.Join(dir, configDirName, githubTokenFileName), []byte(token), os.FileMode(0o600))
+	if err != nil {
+		return fmt.Errorf("saving token file: %w", err)
+	}
+	return nil
+}
+
+func (di *defaultImplementation) readToken() (string, error) {
+	dir, err := os.UserConfigDir()
+	if err != nil {
+		return "", fmt.Errorf("getting user config dir: %w", err)
+	}
+
+	data, err := os.ReadFile(filepath.Join(dir, "slsa", "sourcetool.github.token"))
```

Comment:
use `configDirName` and `githubTokenFileName` globals

Reply:
Nice catch, thanks!